### ISSUE TITLE
Visual Studio instance not found when building python package gpu version when using VS2022 and how I solved it

### ISSUE THREAD

**Issue by 124san** [microsoft/LightGBM#4906]
I tried to install python package gpu version after downloading and installing all the prerequisites, including Visual Studio, Boost and OpenCL. However, it resulted in "Exception Please install Visual Studio or MS Build and all required dependencies first" (screenshot doesnt have any install option because it was for reproducing the error, which the same with install options included):
<img width="1067" alt="Screenshot 2021-12-23 152149" src="https://user-images.githubusercontent.com/27974060/147208095-4d7ffd68-131b-4dcc-89a1-becc459f93ce.png">
Checking the logs gives me something like this:
![image](https://user-images.githubusercontent.com/27974060/147208029-1f94fccb-16e5-477e-808b-d7ec9a2aaddd.png)

The first thing I did after the error is to install 2019 build tools within **Visual Studio 2022**. However, it still gave me the same error after installing. I had no luck with 2017 build tools either.

After a decent amount of googling, I found [this post](https://stackoverflow.com/questions/66376188/cmake-cant-find-instance-of-visual-studio-15-2017-generator-but-works-fine-with) on stackoverflow. The accepted answer suggested installing **the entire "Visual Studio Build Tools 2017" which include MSBuild.** After downloading and installing 2019 build tools, the build command for gpu version finally succeeded. Here is the command I used, with 1.78.0 boost: `pip install lightgbm --install-option=--gpu --install-option="--opencl-include-dir=C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.5/include" --install-option="--opencl-library=C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.5/lib/x64/OpenCL.lib" --install-option="--boost-root=C:/local/boost_1_78_0" --install-option="--boost-librarydir=C:/local/boost_1_78_0/lib64-msvc-14.2"`.

TLDR: Python LightGBM gpu version build process does not pick up Visual Studio 2022 version, even if it has older version of build tools inside. You have to install older Visual Studio or Build Tools **seperately**.

**Comment by StrikerRUS** [comment]
@124san Thanks for reporting this issue!

Support for Visual Studio 2022 was added a few days ago (#4889) and next release of LightGBM will respect it correctly. For now you can install nightly builds: https://lightgbm.readthedocs.io/en/latest/Installation-Guide.html#nightly-builds.

Just a note: I'm glad that GPU version finely worked with Boost 1.78 and CUDA 11.5 (we are not testing this combination of dependencies on our CI)!

**Comment by jameslamb** [comment]
One other note... @124san in the future, if possible, please post error messages and logs as text instead of screenshots. That makes it possible for others facing similar issues to find these discussions from search engines.

**Comment by 124san** [comment]
I didn't think about search engine, thank you for letting me know! I'll
post text next time.

**Comment by 124san** [comment]
> @124san Thanks for reporting this issue!
> 
> Support for Visual Studio 2022 was added a few days ago (#4889) and next release of LightGBM will respect it correctly. For now you can install nightly builds: https://lightgbm.readthedocs.io/en/latest/Installation-Guide.html#nightly-builds.
> 
> Just a note: I'm glad that GPU version finely worked with Boost 1.78 and CUDA 11.5 (we are not testing this combination of dependencies on our CI)!

I'm glad that it works with newest Boost and CUDA! Also, thanks for letting me know about the support for VS2022.
